src/hb-subset-plan.cc

```diff
@@ -91,94 +91,102 @@ _remap_indexes (const hb_set_t *indexes,
 typedef void (*layout_collect_func_t) (hb_face_t *face, hb_tag_t table_tag, const hb_tag_t *scripts, const hb_tag_t *languages, const hb_tag_t *features, hb_set_t *lookup_indexes /* OUT */);
 
 
-template <typename T>
-static void _collect_layout_indices (hb_face_t		  *face,
-                                     const T&              table,
-                                     const hb_set_t	  *layout_features_to_retain,
-                                     layout_collect_func_t layout_collect_func,
-                                     hb_set_t		  *indices /* OUT */)
+/*
+ * Removes all tags from 'tags' that are not in filter. Additionally eliminates any duplicates.
+ * Returns true if anything was removed (not including duplicates).
+ */
+static bool _filter_tag_list(hb_vector_t<hb_tag_t>* tags, /* IN/OUT */
+                             const hb_set_t* filter)
 {
-  hb_vector_t<hb_tag_t> features;
-  if (!features.alloc (table.get_feature_count () + 1))
-    return;
+  hb_vector_t<hb_tag_t> out;
+  out.alloc (tags->get_size());
+
+  bool removed = false;
+  hb_set_t visited;
 
-  hb_set_t visited_features;
-  bool retain_all_features = true;
-  for (unsigned i = 0; i < table.get_feature_count (); i++)
+  for (hb_tag_t tag : *tags)
   {
-    hb_tag_t tag = table.get_feature_tag (i);
     if (!tag) continue;
-    if (!layout_features_to_retain->has (tag))
+    if (visited.has (tag)) continue;
+
+    if (!filter->has (tag))
     {
-      retain_all_features = false;
+      removed = true;
       continue;
     }
 
-    if (visited_features.has (tag))
-      continue;
-
-    features.push (tag);
-    visited_features.add (tag);
+    visited.add (tag);
+    out.push (tag);
   }
 
-  if (!features)
+  hb_swap (out, *tags);
+  return removed;
+}
+
+template <typename T>
+static void _collect_layout_indices (hb_subset_plan_t     *plan,
+                                     const T&              table,
+                                     layout_collect_func_t layout_collect_func,
+                                     hb_set_t		  *indices /* OUT */)
+{
+  unsigned num_features = table.get_feature_count () + 1;
+  hb_vector_t<hb_tag_t> features;
+  if (!plan->check_success (features.resize (num_features))) return;
+  table.get_feature_tags (0, &num_features, features.arrayZ);
+  features.resize (num_features);
+  bool retain_all_features = !_filter_tag_list (&features, plan->layout_features);
+
+  unsigned num_scripts = table.get_script_count () + 1;
```

Comment:
What's the +1 for? And why a second resize below?

Reply:
That was meant to add extra space for the null terminating entry, but this was the wrong place for it. I've moved the +1 into the filter method which produces the vector that actually has the null entry added to the end.